add `bs.inline` section (#23)
melange-re/documentation-site: docs/communicate-with-javascript.md

| None -> Js.log "non-node environment"
```

## Inlining constant values

Some JavaScript idioms require special constants to be inlined since they serve
as de-facto directives for bundlers. A common example is `process.env.NODE_ENV`:

```js

> if(process.env.NODE_ENV !== 'production') {
>    // Development-only code
> }

Reviewer: Suggesting same style and formatting as generated output for consistency.

```suggestion
if (process.env.NODE_ENV !== "production") {
  // Development-only code
}
```

```

becomes:

```js

> if('development' !== 'production') {
>    // Development-only code
> }

Reviewer: ```suggestion
if ("development" !== "production") {
  // Development-only code
}
```

```

In this case, bundlers such as Webpack can tell that the `if` statement always
evaluates to a specific branch and eliminate the others.

> Melange provides the `[@bs.inline]` construct to achieve the same goal in

Reviewer: I have been trying (not sure if I was consistent enough) to refer to attributes by just their name, without the syntactic `[]` brackets. Wdyt?

```suggestion
Melange provides the `bs.inline` attribute to achieve the same goal in
```

> generated JavaScript. Let's look at an example:

Reviewer: ```suggestion
generated JavaScript. Let’s look at an example:
```

https://english.stackexchange.com/a/36048

Reviewer: I think it's Option + Shift + ] on macos us keyboards. In any case, I can make a find/replace when we are done with v1.

```ocaml
external node_env : string = "NODE_ENV" [@@bs.val] [@@bs.scope "process", "env"]

> (* This variable will be used in the comparison below. *)

Reviewer: Is this comment necessary? The value is used in the line below.

Author: It's about the generated code. Should we make it more explicit?

let development = "development"
let () = if node_env <> development then Js.log "Only in Production"

(* The value of this variable will be inlined in the final output. The variable

>    won't be present in the JavaScript output. *)

Reviewer: Also unsure about this comment, as it is mentioned in the "As we can see" paragraph below.

Author: OK I'll remove the comments.

let development_inline = "development" [@@bs.inline]
let () = if node_env <> development_inline then Js.log "Only in Production"
```

As we can see in the generated JavaScript presented below:

- the `development` variable is emitted
    - it gets used as a variable `process.env.NODE_ENV !== development` in the
      `if` statement

> - the `development_inline` variable isn't present in the final output

Reviewer: ```suggestion
- the `development_inline` variable isn’t present in the final output
```

    - its value is inlined in the `if` statement: `process.env.NODE_ENV !== "development"`

```js

> // Generated by Melange
> 'use strict';
> 
> 
> var development = "development";
> 
> if (process.env.NODE_ENV !== development) {
>   console.log("Only in Production");
> }
> 
> if (process.env.NODE_ENV !== "development") {
>   console.log("Only in Production");
> }
> 
> exports.development = development;
> /*  Not a pure module */

Reviewer: We can skip the first lines and exports from the generated JS, to avoid some noise (I don't think they add much when presented in the docs):

```suggestion
var development = "development";

if (process.env.NODE_ENV !== development) {
  console.log("Only in Production");
}

if (process.env.NODE_ENV !== "development") {
  console.log("Only in Production");
}
```

var v = Path.join("a", 2);
```

> ### Bing to a polymorphic function

Reviewer: 🤣 🙏

You can also use any JSON literal by passing a quoted string to `bs.as`: `bs.as
{json|true|json}` or `bs.as {json|{"name": "John"}|json}`.

> Sometimes, passing a constant argument to an external function is not enough.

Reviewer: "is not enough" for what? I am not sure I get the connection between this "constant values as arguments" section and the `bs.inline` attribute 🤔

Author: this connection probably doesn't make sense indeed. I'll remove.

> If you need to inline constant values, check [`[@bs.inline]`](#bs-inline)

Reviewer: Same suggestion as above about brackets:

```suggestion
If you need to inline constant values, check the [`bs.inline`](#bs-inline) attribute.
```
